### PR TITLE
[FLOC-1461] [FLOC-1462] [FLOC-1383] Add systemd units

### PR DESCRIPTION
**Pull request by tomprince**
https://clusterhq.atlassian.net/browse/FLOC-1461
https://clusterhq.atlassian.net/browse/FLOC-1462

This is based on https://github.com/ClusterHQ/flocker/compare/control-firewall-FLOC-1466...systemd-units-FLOC-1461 to get the `package-files` work done there, as well as to expand the documentation sketch there.

This should also eventually address
https://clusterhq.atlassian.net/browse/FLOC-1383

### CODE REVIEW

File: admin/package-files/flocker-control-agent.socket [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[Unit]
```
Comment (itamarst) [inline admin/package-files/flocker-control-agent.socket]: I suggest having a `systemd` folder for systemd-specific files.

File: admin/package-files/flocker-control.service [hunk 2]
```diff
@@ -0,0 +1,17 @@
+[Unit]
+Description=Flocker Control Service
+
+[Service]
+ExecStart=/usr/bin/flocker-control -p systemd:0 -a systemd:1
```
Comment (itamarst) [inline admin/package-files/flocker-control.service]: Can we actually do TLS off of systemd-passed-sockets like this? As first approximate I would guess yes, but if not then maybe we can't use this systemd feature.

File: admin/package-files/flocker-zfs-agent.service [hunk 3]
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=Flocker ZFS Agent
+
+[Service]
+ExecStart=/usr/bin/flocker-zfs-agent $FLOCKER_NODE_NAME $FLOCKER_CONTROL_NODE
```
Comment (itamarst) [inline admin/package-files/flocker-zfs-agent.service]: So docs will say "edit `/etc/sysconfig/flocker-zfs-agent` and set these options"? I suggest maybe calling it (and other related files correspondingly) `/etc/sysconfig/flocker-agent`, we're probably going to replace the ZFS-specific agent with a generic agent that figures out if it should do ZFS or EBS or what.

**Comment by itamarst** [comment]
Looks good, with comments above for questions/suggestions.

File: admin/package-files/flocker-control.service (hunk 2 above)
Comment (tomprince) [inline admin/package-files/flocker-control.service]: https://twistedmatrix.com/documents/current/api/twisted.protocols.tls.TLSMemoryBIOFactory.html

File: docs/advanced/api.rst [hunk 4]
```diff
@@ -14,12 +14,24 @@ Installation
...
+
+The control service needs to accessible remotely.
+To configure firewalld to allow access to the control service REST API, and for agent connections,
 
 .. task:: open_control_firewall
 
+(On AWS, an external firewall is used instead, which will need to be configured similarity).
 For more details on configuring the firewall, see Fedora's `FirewallD documentation <https://fedoraproject.org/wiki/FirewallD>`_.
 
+To start the ZFS agent on a node, (where ``${CONTROL_NODE}`` is the address of the control node, and ``${NODE_NAME}`` is the name of the node being configured.:
```
Comment (adamtheturtle) [inline docs/advanced/api.rst]: Unmatched parenthesis.

File: admin/acceptance.py [hunk 5]
```diff
@@ -68,11 +69,15 @@ def extend_environ(**kwargs):
     return env
 
 
-def run_tests(nodes, trial_args):
+def run_tests(nodes, control_node, agent_nodes, trial_args):
     """
     Run the acceptances tests.
```
Comment (adamtheturtle) [inline admin/acceptance.py]: (existing) typo "acceptances tests" " acceptance tests".

File: flocker/acceptance/test_api.py [hunk 6]
```diff
@@ -252,84 +184,41 @@ def create_container(self, properties):
...
-    ``flocker-zfs-agent`` is started on every node.
-
-    The processes will be killed after each test.
+    Build a ``Cluster`` instance with at least ``node_count`` nodes.
 
     :param TestCase test_case: The test case instance on which to register
         cleanup operations.
-    :param list node_address: The IPv4 addresses of the nodes in the cluster.
-    :returns: A ``Cluster`` instance.
+    :param list node_count: The number of nodes to request in the cluster.
```
Comment (adamtheturtle) [inline flocker/acceptance/test_api.py]: This is an integer, not a list.

File: flocker/acceptance/test_api.py [hunk 7]
```diff
@@ -252,84 +184,41 @@ def create_container(self, properties):
...
-        nodes.append(node)
+    control_node = environ.get('FLOCKER_ACCEPTANCE_CONTROL_NODE')
 
-    return Cluster(control_service=control_service, nodes=nodes)
+    if control_node is None:
+        raise SkipTest(
+            "Set acceptance testing control node IP address using the " +
+            "FLOCKER_ACCEPTANCE_CONTROL_NODE environment variable.")
 
+    agent_nodes_env_var = environ.get('FLOCKER_ACCEPTANCE_AGENT_NODEs')
```
Comment (adamtheturtle) [inline flocker/acceptance/test_api.py]: `FLOCKER_ACCEPTANCE_AGENT_NODEs` > `FLOCKER_ACCEPTANCE_AGENT_NODES`

File: flocker/provision/_install.py [hunk 8]
```diff
@@ -200,6 +210,29 @@ def task_open_control_firewall():
...
 
+AGENT_CONFIG = """\
+FLOCKER_NODE_NAME = %(node_name)s
+FLOCKER_CONTROL_NODE = %(control_node)s
+"""
+
+
+def task_enable_flocker_agent(node_name, control_node):
+    """
+    Configure and eanble flocker-agent.
```
Comment (adamtheturtle) [inline flocker/provision/_install.py]: Typo `eanble` > `enable`.

File: flocker/provision/_install.py [hunk 9]
```diff
@@ -200,6 +210,29 @@ def task_open_control_firewall():
     ])
 
 
+AGENT_CONFIG = """\
+FLOCKER_NODE_NAME = %(node_name)s
+FLOCKER_CONTROL_NODE = %(control_node)s
+"""
+
+
+def task_enable_flocker_agent(node_name, control_node):
```
Comment (adamtheturtle) [inline flocker/provision/_install.py]: Document parameters.

File: flocker/acceptance/test_api.py [hunk 10]
```diff
@@ -4,112 +4,44 @@
...
     testing.
 
-    :param RemoteService control_service: The remotely running
-        ``flocker-control`` process.
+    :param Node control_node: The node running the ``flocker-control``
+        service.
     :param list nodes: The ``Node`` s in this cluster.
     """
+    control_node = field(type=Node)
+    nodes = field()  # Plist(Node)
```
Comment (adamtheturtle) [inline flocker/acceptance/test_api.py]: I'm not sure what to make of this comment.

File: flocker/provision/_install.py [hunk 11]
```diff
@@ -190,6 +190,16 @@ def task_disable_firewall():
          'FORWARD', '0', '-j', 'ACCEPT'])
 
 
+def task_enable_flocker_control():
+    """
+    Eanble flocker-control service.
```
Comment (adamtheturtle) [inline flocker/provision/_install.py]: Typo "Eanble" > "Enable"

File: flocker/provision/_rackspace.py [hunk 12]
```diff
@@ -30,6 +30,7 @@ def provision_rackspace(node, package_source, distribution):
             distribution=node.distribution,
         )
         + task_disable_firewall()
+        # This should be part of ._install.configure_cluster
```
Comment (adamtheturtle) [inline flocker/provision/_rackspace.py]: Do this or create an issue for it or remove the comment.

File: flocker/provision/_sphinx.py [hunk 13]
```diff
@@ -70,7 +79,15 @@ def run(self):
             except KeyError:
                 raise self.error("task: %s not supported"
                                  % (type(command).__name__,))
-            lines += ['   %s' % (line,) for line in handler(command)]
+            # handler can return either a string or a list.  If it is a string,
```
Comment (adamtheturtle) [inline flocker/provision/_sphinx.py]: "If it is a list" surely? (also, "if it is" > "if it returns").

File: admin/acceptance.py [hunk 14]
```diff
@@ -200,6 +208,8 @@ def start_nodes(self):
                 raise
 
             self.nodes.append(node)
+
+            check_safe_call(['ssh-keygen', '-R', node.address])
```
Comment (adamtheturtle) [inline admin/acceptance.py]: I had to use `man` to find what this does.

```
     -R hostname
             Removes all keys belonging to hostname from a known_hosts file.  This option is useful to delete hashed hosts (see
             the -H option above).
```

A comment would be helpful.

File: admin/acceptance.py [hunk 15]
```diff
@@ -365,7 +375,11 @@ def main(args, base_path, top_level):
 
     try:
         nodes = runner.start_nodes()
-        result = run_tests(nodes, options['trial-args'])
+        configure_cluster(control_node=nodes[0], agent_nodes=nodes)
+        result = run_tests(
+            nodes=nodes,
+            control_node=nodes[0], agent_nodes=nodes,
```
Comment (adamtheturtle) [inline admin/acceptance.py]: Is it necessary, or meant to be obvious that the control node must be one of the agent nodes. Sorry if I'm missing some documentation, but if I'm not please make that clear if it is necessary.

File: docs/advanced/api.rst [hunk 16]
```diff
@@ -14,12 +14,24 @@ Installation
 Fedora/CentOS
 -------------
 
-To allow remote access to the control agent REST API, and for agent connections,
+To enable Flocker control service.
```
Comment (adamtheturtle) [inline docs/advanced/api.rst]: Elsewhere in the docs this is "the control service" so perhaps that or "the Flocker control service".

Comment (adamtheturtle) [inline docs/advanced/api.rst]: Also this should end with a colon, and perhaps explain a little about why this is necessary.

File: docs/advanced/api.rst [hunk 17]
```diff
@@ -14,12 +14,24 @@ Installation
...
+
+.. task:: enable_flocker_control
+   :prompt: [root@control-node]#
+
+The control service needs to accessible remotely.
+To configure firewalld to allow access to the control service REST API, and for agent connections,
 
 .. task:: open_control_firewall
 
+(On AWS, an external firewall is used instead, which will need to be configured similarity).
```
Comment (adamtheturtle) [inline docs/advanced/api.rst]: Is there a follow up issue to document this configuration?

If not, please make one.
If so, please either link to that here (in a comment) or mention this section on that issue as something to change.

File: docs/advanced/api.rst [hunk 18]
```diff
@@ -14,12 +14,24 @@ Installation
...
 -------------
 
-To allow remote access to the control agent REST API, and for agent connections,
+To enable Flocker control service.
+
+.. task:: enable_flocker_control
+   :prompt: [root@control-node]#
+
+The control service needs to accessible remotely.
+To configure firewalld to allow access to the control service REST API, and for agent connections,
```
Comment (adamtheturtle) [inline docs/advanced/api.rst]: :question: It seems from the link below that this should be `FirewallD`.

File: flocker/acceptance/test_api.py [hunk 19]
```diff
@@ -4,112 +4,44 @@
...
-def close(process):
     """
-    Kill a process.
-
-    :param subprocess.Popen process: The process to be killed.
-    """
-    process.stdin.close()
-    kill(process.pid, SIGINT)
-
+    address = field(type=bytes)
```
Comment (adamtheturtle) [inline flocker/acceptance/test_api.py]: :question: Would it perhaps be better to use an `IPv4Address` instead of `bytes`.

File: docs/advanced/api.rst [hunk 20]
```diff
@@ -14,12 +14,24 @@ Installation
...
 
 .. task:: open_control_firewall
 
+(On AWS, an external firewall is used instead, which will need to be configured similarity).
 For more details on configuring the firewall, see Fedora's `FirewallD documentation <https://fedoraproject.org/wiki/FirewallD>`_.
 
+To start the ZFS agent on a node, (where ``${CONTROL_NODE}`` is the address of the control node, and ``${NODE_NAME}`` is the name of the node being configured).:
```
Comment (adamtheturtle) [inline docs/advanced/api.rst]: Is it important that this is specifically a ZFS agent? Is it always going to be?

If it is not necessary to specify ZFS then that should be removed.

File: docs/advanced/api.rst [hunk 21]
```diff
@@ -14,12 +14,24 @@ Installation
 Fedora/CentOS
 -------------
 
-To allow remote access to the control agent REST API, and for agent connections,
+To enable Flocker control service.
+
+.. task:: enable_flocker_control
```
Comment (adamtheturtle) [inline docs/advanced/api.rst]: Because of https://github.com/sbrunner/sphinx-prompt/issues/2 this is broken in at least Safari and Chrome. That issue did not affect us much because previously it was only AFAIK a minor problem in the release process docs. See the screenshot:
screenshot 2015-03-16 14 52 16

This really confused me reading the rendered docs. That tells me that there should be prose explaining where these commands are run. There are a few options:
* Live with the rendering problem and hope that upstream fixes it. I don't like this option.
* Not use `sphinx-prompt` here.
* Submit a fix upstream.

File: admin/package-files/systemd/flocker-agent.service [hunk 22]
```diff
@@ -0,0 +1,11 @@
+[Unit]
+Description=Flocker Node Agent
+
+[Service]
+ExecStart=/usr/bin/flocker-zfs-agent $FLOCKER_NODE_NAME $FLOCKER_CONTROL_NODE
```
Comment (adamtheturtle) [inline admin/package-files/systemd/flocker-agent.service]: @itamarst made a suggestion to change this at https://github.com/ClusterHQ/flocker/pull/1179#discussion_r25618137. Is there a reason not to?

Comment (tomprince) [inline admin/package-files/systemd/flocker-agent.service]: I don't think changing the name of the binary is in scope for this ticket.

**Comment by adamtheturtle** [comment]
Thanks @tomprince , please address the above comments and resubmit.

File: docs/advanced/api.rst (hunk 17 above)
Comment (adamtheturtle) [inline docs/advanced/api.rst]: Just AWS? Or also other cloud providers?

Comment (tomprince) [inline docs/advanced/api.rst]: This is just AWS, the other providers don't have an external firewall. I'm not going to file a issues for this: it isn't clear to me that this belongs in our documentation (we can't document all the details about how to use every cloud provider). If it is in scope, it can be filed as part of a specific user story.

File: docs/advanced/api.rst (hunk 16 above)
Comment (tomprince) [inline docs/advanced/api.rst]: > perhaps explain a little about why this is necessary.

I'm going to say that this is outside the scope of this issue. This document isn't close to ready for an end-user to take and start using. When those documents are written, this can be expanded as necessary.

File: admin/acceptance.py (hunk 15 above)
Comment (tomprince) [inline admin/acceptance.py]: It isn't necessary. For the tests, it doesn't seem to make a lot of sense to spin up an extra node for it, at least at this point.

File: docs/advanced/api.rst (hunk 21 above)
Comment (tomprince) [inline docs/advanced/api.rst]: The installation instructions already use `sphinx-prompt`, so I'm not going to consider this a blocker for this issue.

Comment (tomprince) [inline docs/advanced/api.rst]: https://clusterhq.atlassian.net/browse/FLOC-1548

File: docs/advanced/api.rst (hunk 18 above)
Comment (tomprince) [inline docs/advanced/api.rst]: If you follow the link, it looks like most of the uses there are uncapitalized. So I uncapitalized the link as well, instead.

File: flocker/acceptance/test_api.py (hunk 19 above)
Comment (tomprince) [inline flocker/acceptance/test_api.py]: This could also be a hostname.

File: admin/package-files/systemd/flocker-control-agent.socket [hunk 23]
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=Flocker Control Agent Scoket
```
Comment (itamarst) [inline admin/package-files/systemd/flocker-control-agent.socket]: "Scoket"

**Comment by itamarst** [comment]
Looks good, one typo then merge.

**Comment by itamarst** [comment]
Oh, and fix failing tests :grin:

**Comment by adamtheturtle** [comment]
https://clusterhq.atlassian.net/browse/FLOC-1474 can be closed by this too, right?
